### PR TITLE
Map \x01ACTION to the m.emote msgtype

### LINKED ISSUE #36: \x01ACTION needs to be mapped to msgtype m.emote
```
{
  "origin_server_ts": 1486082822216,
  "sender": "@madduck:madduck.net",
  "event_id": "$1486082822202NYBgM:madduck.net",
  "unsigned": {
    "age": 178
  },
  "content": {
    "body": "\u0001ACTION uses /me to exemplify\u0001",
    "msgtype": "m.text"
  },
  "type": "m.room.message",
  "room_id": "!qYuPEjvrchmJnELuQl:jki.re"
}
```

This should be msgtype m.emote instead :)

### PR DESCRIPTION
**Pull request by vmiklos**
Fixes <https://github.com/matrix-org/matrix-ircd/issues/36>.

### CODE REVIEW

**Comment by vmiklos** [comment]
Ah, forgot the sign-off.

**Comment by vmiklos** [comment]
Forgot to run cargo test, let me fix that up.

File: src/bridge/mod.rs [hunk 1]
```diff
@@ -52,6 +52,20 @@ pub struct Bridge<IS: AsyncRead + AsyncWrite + 'static> {
     joining_map: BTreeMap<String, String>,
 }
 
+fn handle_privmsg_action(text: &str) -> (String, String) {
```
Comment (jplatte) [inline src/bridge/mod.rs]: I find this harder to read than necessary. How about this instead:

```rust
let emote_content = text.strip_prefix(..).and_then(|noprefix| noprefix.strip_suffix(..));
match emote_content {
    Some(value) => ("m.emote", value.into()),
    None => ("m.text", text),
}
```

(and also make `text` a `String` so we don't needlessly copy it)

Comment (vmiklos) [reply src/bridge/mod.rs]: Done, now using .and_then() to simplify the code.

Comment (vmiklos) [reply src/bridge/mod.rs]: And now switched text to be a String. :-)

**Comment by jplatte** [comment]
> Ah, forgot the sign-off.

You don't need to rewrite your commit and force-push for adding a sign-off, it can simply be in a comment / the PR description.

File: src/bridge/mod.rs [hunk 2]
```diff
@@ -52,6 +52,17 @@ pub struct Bridge<IS: AsyncRead + AsyncWrite + 'static> {
     joining_map: BTreeMap<String, String>,
 }
 
+fn handle_privmsg_action(text: &String) -> (String, String) {
```
Comment (jplatte) [inline src/bridge/mod.rs]:
```suggestion
fn handle_privmsg_action(text: String) -> (&'static str, String) {
```

File: src/bridge/mod.rs [hunk 3]
```diff
@@ -52,6 +52,17 @@ pub struct Bridge<IS: AsyncRead + AsyncWrite + 'static> {
...
+    // Example: '\x01ACTION foo\x01'.
+    let emote_content = text
+        .strip_prefix("\x01ACTION")
+        .and_then(|noprefix| noprefix.strip_suffix("\x01"));
+    match emote_content {
+        Some(value) => ("m.emote".to_string(), value.into()),
```
Comment (jplatte) [inline src/bridge/mod.rs]:
```suggestion
        Some(value) => ("m.emote", value.into()),
```

File: src/bridge/mod.rs [hunk 4]
```diff
@@ -52,6 +52,17 @@ pub struct Bridge<IS: AsyncRead + AsyncWrite + 'static> {
...
+        None => ("m.text".to_string(), text.to_string()),
```
Comment (jplatte) [inline src/bridge/mod.rs]:
```suggestion
        None => ("m.text", text),
```

File: src/bridge/mod.rs [hunk 5]
```diff
@@ -118,9 +129,10 @@ impl<IS: AsyncRead + AsyncWrite + 'static> Bridge<IS> {
             IrcCommand::PrivMsg { channel, text } => {
                 if let Some(room_id) = self.mappings.channel_to_room_id(&channel) {
                     info!(self.ctx.logger, "Got msg"; "channel" => channel.as_str(), "room_id" => room_id.as_str());
+                    let (msgtype, body) = handle_privmsg_action(&text);
```
Comment (jplatte) [inline src/bridge/mod.rs]:
```suggestion
                    let (msgtype, body) = handle_privmsg_action(text);
```

**Review by jplatte — COMMENTED** [review]
Now it's `&String` (not `String`) and still copied. Maybe something about what I was imagining doesn't actually work, or I wasn't clear enough. I've added change suggestions for what I think should work, let me know if it doesn't.

File: src/bridge/mod.rs (hunk 2 above)
Comment (vmiklos) [reply src/bridge/mod.rs]: > Now it's `&String` (not `String`) and still copied. Maybe something about what I was imagining doesn't actually work, or I wasn't clear enough. I've added change suggestions for what I think should work, let me know if it doesn't.

Thanks, I just didn't consider `&'static str` when you mentioned no-copying. :-) Hopefully it now addresses your concern.

**Comment by jplatte** [comment]
Thanks!

**Pull request merged**